contracts/src/L2/ETHRegistrar.sol

```diff
@@ -151,60 +171,67 @@ contract ETHRegistrar is IETHRegistrar, EnhancedAccessControl {
-        uint256 totalPrice = checkPrice(name, duration);
-
+        uint64 duration,
+        address token
+    ) external returns (uint256 tokenId) {
+        // CHECKS: Validate commitment and get pricing (external calls for validation only)
         _consumeCommitment(name, duration, makeCommitment(name, owner, secret, address(subregistry), resolver, duration));
-
         uint64 expiry = uint64(block.timestamp) + duration;
-        tokenId = registry.register(name, owner, subregistry, resolver, REGISTRATION_ROLE_BITMAP, expiry);
-
-        if (msg.value > totalPrice) {
-            payable(msg.sender).transfer(msg.value - totalPrice);
+        // Get USD pricing breakdown
+        ITokenPriceOracle.Price memory usdPrice = prices.price(name, expiry, duration);
+        
+        // Convert to token amount for payment and handle transfer
+        {
+            ITokenPriceOracle tokenOracle = prices;
+            uint256 tokenAmount = tokenOracle.priceInToken(name, expiry, duration, token);
+            // EFFECTS: Handle payment BEFORE state changes
+            IERC20(token).safeTransferFrom(msg.sender, beneficiary, tokenAmount);
         }
 
-        emit NameRegistered(name, owner, subregistry, resolver, duration, tokenId);
+        // INTERACTIONS: Register name only after successful payment
+        tokenId = registry.register(name, owner, subregistry, resolver, REGISTRATION_ROLE_BITMAP, expiry);
+        
+        emit NameRegistered(name, owner, subregistry, resolver, duration, tokenId, usdPrice.base, usdPrice.premium);
     }
 
     /**
-     * @dev Renew a name.
+     * @dev Renew a name with ERC20 token payment.
      * @param name The name to renew.
      * @param duration The duration of the renewal.
+     * @param token The ERC20 token address for payment.
      */
-    function renew(string calldata name, uint64 duration) external payable {
-        uint256 totalPrice = checkPrice(name, duration);
-
+    function renew(string calldata name, uint64 duration, address token) external {
+        // CHECKS: Get current data and validate pricing
         (uint256 tokenId, uint64 expiry, ) = registry.getNameData(name);
-
-        registry.renew(tokenId, expiry + duration);
-
-        if (msg.value > totalPrice) {
-            payable(msg.sender).transfer(msg.value - totalPrice);
+        
+        // Check for overflow before any state changes
+        if (expiry > type(uint64).max - duration) {
+            revert DurationOverflow(expiry, duration);
         }
-
-        uint64 newExpiry = registry.getExpiry(tokenId);
-
-        emit NameRenewed(name, duration, tokenId, newExpiry);
+        uint64 newExpiry = expiry + duration;
+        
+        // Get USD pricing breakdown
+        ITokenPriceOracle.Price memory usdPrice = prices.price(name, uint256(expiry), duration);
+        
+        // Convert to token amount for payment and handle transfer
+        {
+            ITokenPriceOracle tokenOracle = prices;
```

Comment:
Redundant. Can use it directly - perhaps we should just call the member variable to tokenOracle instead of prices. 

Actually let's call it `tokenPriceOracle` instead of prices.